src/perf/bin/appmain.cpp

```diff
@@ -0,0 +1,204 @@
+    _In_ int argc,
+    _In_reads_(argc) _Null_terminated_ char* argv[],
+    _In_ bool KeyboardWait
+    ) {
+    QUIC_EVENT StopEvent;
+    QuicEventInitialize(&StopEvent, true, false);
+
+    int RetVal = QuicMainStart(argc, argv, StopEvent);
+    if (RetVal != 0) {
+        return RetVal;
+    }
+    printf("Ready For Connections!\n\n");
+    fflush(stdout);
+    if (KeyboardWait) {
+        printf("Press enter to exit\n");
+        getchar();
+        QuicEventSet(StopEvent);
+    }
+    RetVal = QuicMainStop(0);
+    QuicEventUninitialize(StopEvent);
+    return RetVal;
+}
+
+#ifdef _WIN32
+
+int
+QuicKernelMain(
+    _In_ int argc,
+    _In_reads_(argc) _Null_terminated_ char* argv[],
+    _In_ bool KeyboardWait,
+    _In_ QUIC_SEC_CONFIG_PARAMS* SelfSignedParams
+    ) {
+    size_t TotalLength = 0;
+
+    // Get total length
+    for (int i = 0; i < argc; ++i) {
+        //
+        // Length of string, plus null terminator, plus length
+        //
+        TotalLength += strlen(argv[i]) + sizeof(size_t) + 1;
+    }
+
+    TotalLength += sizeof(TotalLength);
+
+    char* Data = static_cast<char*>(QuicAlloc(TotalLength));
+    if (!Data) {
+        printf("Failed to allocate arguments to pass\n");
+        return QUIC_RUN_FAILED_TEST_INITIALIZE;
+    }
+
+    char* DataCurrent = Data;
+
+    QuicCopyMemory(DataCurrent, &TotalLength, sizeof(TotalLength));
+
+    DataCurrent += sizeof(TotalLength);
+
+    for (int i = 0; i < argc; ++i) {
+        size_t ArgLen = strlen(argv[i]) + 1;
+        QuicCopyMemory(DataCurrent, &ArgLen, sizeof(ArgLen));
+        DataCurrent += sizeof(ArgLen);
+        QuicCopyMemory(DataCurrent, argv[i], ArgLen);
+        DataCurrent += ArgLen;
+        DataCurrent[0] = '\0';
+        ++DataCurrent;
+    }
+
+    QUIC_DBG_ASSERT(DataCurrent == (Data + TotalLength));
+    QUIC_DBG_ASSERT(TotalLength <= UINT_MAX);
+
+    constexpr DWORD OutBufferSize = 1024 * 1000;
+    char* OutBuffer = (char*)QuicAlloc(OutBufferSize); // 1 MB
+    if (!OutBuffer) {
+        printf("Failed to allocate space for output buffer\n");
+        QuicFree(Data);
+        return QUIC_RUN_FAILED_TEST_INITIALIZE;
+    }
+
+    QuicDriverService DriverService;
+    QuicDriverClient DriverClient;
+    DriverService.Initialize();
+    DriverService.Start();
+    DriverClient.Initialize(SelfSignedParams);
+
+    if (!DriverClient.Run(IOCTL_QUIC_RUN_PERF, Data, (uint32_t)TotalLength)) {
+        QuicFree(Data);
+        QuicFree(OutBuffer);
+        return QUIC_RUN_FAILED_TEST_INITIALIZE;
+    }
+    printf("Ready For Connections!\n\n");
+    fflush(stdout);
+
+    DWORD OutBufferWritten = 0;
+    bool RunSuccess =
+        DriverClient.Read(
+            IOCTL_QUIC_READ_DATA,
+            OutBuffer,
+            OutBufferSize,
+            &OutBufferWritten);
+
+    if (RunSuccess) {
+        printf("%s", OutBuffer);
+    }
+
+    QuicFree(Data);
+    QuicFree(OutBuffer);
+
+    return RunSuccess ? QUIC_RUN_SUCCESS : QUIC_RUN_STOP_FAILURE;
```

Comment:
It's not actually stop failure though, right? Also, why even return a status here? Generally, if the caller isn't going to do anything special with a specific error, and just needs to know success/fail, return a bool. If they aren't going to do anything special with a failure, return void. For most of this stuff, you can just print something at the point of failure and bail.

Reply:
It's int because this is what is returned from main. Its a forwarder basically. I can make it bool and then map it on the other side, but thats really not much cleaner.